csharp/src/Ice/Communicator-OutgoingConnectionFactory.cs

```diff
@@ -0,0 +1,193 @@
+// Copyright (c) ZeroC, Inc. All rights reserved.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace ZeroC.Ice
+{
+    public sealed partial class Communicator : IConnectionManager
```

Comment:
Do we need IConnectionManager?

Reply:
This is only used to remove the connection I guess we can have something like

```
if (IsIncoming)
{
   // Remove the connection from the factory, must be called without the connection mutex locked
   _manager?.Remove(this);
}
else
{
   Communicator.Remove(this);
}
```

Reply:
Yes, sounds good. And we should remove IConnectionManager and simply keep a reference on the incoming connection factory for incoming connections.

Reply:
To avoid keeping this interface in the transport API, I end up adding a delegate to the connection that is assigned when the connection is added to the communicator/incoming connection factory maps. 

```
// Delegate used to remove the connection once it has been closed.
internal Action<Connection>? Remove
{
    set
    {
        lock (_mutex)
        {
            // If the connection was closed before the delegate was set execute it immediately otherwise
            // it will be called once the connection is close
            if (_state == ConnectionState.Closed)
            {
                Task.Run(() => value?.Invoke(this));
            }
            else
            {
                _remove = value;
            }
        }
    }
}
```